DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/svd/src/svd.hpp

```diff
@@ -0,0 +1,262 @@
+#ifndef __SVD_HPP__
+#define __SVD_HPP__
+
+#include <sycl/ext/intel/ac_types/ac_complex.hpp>
+#include <sycl/ext/intel/ac_types/ac_int.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/sycl.hpp>
+#include <vector>
+#include "tuple.hpp"
+
+// Headers shared with other oneAPI FPGA samples
+#include "exception_handler.hpp"
+#include "streaming_eigen.hpp"
+#include "streaming_qrd.hpp"
+
+// Headers specific to SVD
+#include "memory_transfers.hpp"
+#include "non_std_convariance.hpp"
+#include "post_process.hpp"
+
+
+// Forward declare the kernel and pipe names
+// (This prevents unwanted name mangling in the optimization report.)
+
+// kernels
+class IDInputMatrixFromDDRToLocalMem;
+class IDInputTransMatrixFromDDRToLocalMem;
+class IDCovarianceMatrixComputation;
+class IDEigenValuesAndVectorsComputation;
+class IDInputMatrixFromDDRToLocalMemPostProcess;
+class IDUSVFromEigens;
+class IDOrthogonalizingU;
+class IDRankDeficientFlagFromLocalMemToDDR;
+class IDUMatrixFromLocalMemToDDR;
+class IDRMatrixPipeTerminate;
+class IDSMatrixFromLocalMemToDDR;
+class IDVMatrixFromLocalMemToDDR;
+
+// pipes
+class IDCovarianceMatrixPipe;
+class IDInputMatrixPipe;
+class IDInputMatrixPipe2;
+class IDEigenValuesPipe;
+class IDEigenVectorsPipe;
+class IDRankDeficientFlagPipe;
+class IDUTempMatrixPipe;
+class IDUMatrixPipe;
+class IDRMatrixPipe;
+class IDSMatrixPipe;
+class IDVMatrixPipe;
+
+template <unsigned rows,            // row size of input matrix
+          unsigned cols,            // col size of input matrix
+          unsigned k_fixed_iteration, // fixed number of QR iteration
+          unsigned k_raw_latency,   // RAW latency for triangular loop
+                                    // optimization in the QR iteration
+          int k_zero_threshold_1e,  // Threshold from which we consider a
+          typename T                // The datatype for the computation
+          >
+double SingularValueDecomposition(
+    std::vector<T> &a_matrix,  // input matrix A (col maj)
+    std::vector<T> &u_matrix,  // left singular vectors, U (col maj)
+    std::vector<T> &s_matrix,  // list of singular values, S
+    std::vector<T> &v_matrix,  // right singular vectors, V (row maj)
+    std::vector<ac_int<1, false>>
+        &rank_deficient_flag,  // Output a flag that is 1 if the input matrix
+                               // is considered rank deficient
+    sycl::queue &q, int matrix_count=1, int repetitions=1) {
+  constexpr int kNumElementsPerDDRBurst = 8;            // total of 256 bit
+  constexpr int kInputMatrixSize = rows * cols;         // row x col
+  constexpr int kEigenVectorsMatrixSize = cols * cols;  // col x col
+  constexpr int kUMatrixSize = rows * rows;
+  constexpr int kRMatrixSize = rows * (rows + 1) / 2;
+
+  // pipes
+  using PipeType = fpga_tools::NTuple<T, kNumElementsPerDDRBurst>;
+
+  using InputMatrixPipe = sycl::ext::intel::pipe<IDInputMatrixPipe, PipeType, 16>;
```

Comment:
Has 16 been chosen randomly?

Reply:
it was for debug reason and should be back to 3